Oracle upgrades, model state simplification

InfluenceFunctional opened the pull request:
Big PR - tested with several nupack oracles, MCMC, random, GFN, energy and heuristic acquisition functions.

- added new nupack oracles (open loop size and distance from target f……old) with an optional stability reweighting
- simplified model state calculation - fast (random with anealing) during the run, and then one at the end using whichever sampler is building queries (often more expensive). Also made it simpler / faster overall: added a toggle for large data loss calculation, reduced samples for random distance calculation, reduced samples for large data loss calculation.
- unified all oracles as being negative everywhere, directly from oracles.py
- unified language mostly everywhere for 'scores' vs 'energies' - for now, if it comes out of an oracle, it's an 'energy', (this is the value we are trying to optimize, whether or not it's actually 'energy' i.e. it could be the number of hairpins) and if it's an output of an acquisition function, it's a 'score' - i.e. may be mixed with uncertainty
- various small cleanups
- obviated oracle.py - we can now run oracles directly in AptamerSeq by feeding it oracle.score (for from oracles import Oracle, oracle = Oracle(config)), instead of calling specific functions
- adjusted gflownet.py for sampling from oracle - some tweaks still to be made here and in config

InfluenceFunctional committed:
- added new nupack oracles (open loop size and distance from target fold) with an optional stability reweighting
- simplified model state calculation - fast (random with anealing) during the run, and then one at the end using whichever sampler is building queries (often more expensive). Also made it simpler / faster overall: added a toggle for large data loss calculation, reduced samples for random distance calculation, reduced samples for large data loss calculation.
- unified all oracles as being negative everywhere, directly from oracles.py
- unified language mostly everywhere for 'scores' vs 'energies' - for now, if it comes out of an oracle, it's an 'energy', (this is the value we are trying to optimize, whether or not it's actually 'energy' i.e. it could be the number of hairpins) and if it's an output of an acquisition function, it's a 'score' - i.e. may be mixed with uncertainty
- various small cleanups
- obviated oracle.py - we can now run oracles directly in AptamerSeq by feeding it oracle.score (for from oracles import Oracle, oracle = Oracle(config)), instead of calling specific functions
- adjusted gflownet.py for sampling from oracle - some tweaks still to be made here and in config

alexhernandezgarcia committed:
black

alexhernandezgarcia committed:
remove commented code

alexhernandezgarcia committed:
oracle_func <- oracleFunc to comply with PEP 8

alexhernandezgarcia committed:
oracle_func <- oracleFunc to comply with PEP 8

alexhernandezgarcia committed:
delete commented code

alexhernandezgarcia commented on aptamers.py:
```diff
@@ -43,18 +43,18 @@ class AptamerSeq:
+            debug=False,
+            reward_beta=1,
+            env_id=None,
+            oracleFunc=None,
```
I will be changing `oracle_func` <- `oracleFunc` and similar cases at least in `gflownet.py` and `aptamers.py` ("my" scripts) to comply with [PEP 8](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names) and keep consistency.

InfluenceFunctional replied:
Oops, this was just me being thoughtless about it. Of course go ahead.

alexhernandezgarcia commented on aptamers.py:
```diff
@@ -43,18 +43,18 @@ class AptamerSeq:
-        debug=False,
-        reward_beta=1,
-        env_id=None,
+            self,
+            max_seq_length=42,
+            min_seq_length=1,
+            nalphabet=4,
+            min_word_len=1,
+            max_word_len=1,
+            proxy=None,
+            allow_backward=False,
+            debug=False,
+            reward_beta=1,
+            env_id=None,
+            oracleFunc=None,
```
There's extra white spaces here for some reasons. No problem because I will run `black` on "my" scripts and it will take care of it automatically :)

InfluenceFunctional replied:
I go with what my IDE tells me - it must be on a different spacing setting

alexhernandezgarcia commented on gflownet.py:
```diff
@@ -41,7 +41,7 @@ def add_args(parser):
         "-y",
         "--yaml_config",
-        default=None,
+        default='config/mk_test_defaults_gfn_only.yml',
```
Do we prefer to have one specific config file, or rather `None`? (not sure myself) If the former, I'd point to the most generic possible config.

InfluenceFunctional replied:
agreed to latter, should be pointed to generic config. I was talking to Danny and we thought we could construct such a config at next meeting

DannySalem commented:
We should just commit a black config to the repo so we can all use the same one.

alexhernandezgarcia committed:
black

alexhernandezgarcia committed:
add return to add_bool_arg to make the effect a bit more explicit

alexhernandezgarcia commented on utils.py:
```diff
@@ -930,3 +915,32 @@ def convert(v):
+    group.add_argument('--' + name, dest=name, action = 'store_true')
+    group.add_argument('--no-' + name, dest=name, action = 'store_false')
+    parser.set_defaults(**{name:default})
+    return parser
```
I have added `return parser` (and updated the calls to the function) to make the effect of the call a bit more explicit. Otherwise one may wonder what will change inside the function.

InfluenceFunctional committed:
- switched proxy to mse loss
- added seeds to random sampler and annealer
- fixed error in model state reporting)
- added proxy history length for convergence criteria
- further aligned score :: energy variable naming

InfluenceFunctional committed:
Merge remote-tracking branch 'origin/mk_al_dev' into mk_al_dev

InfluenceFunctional committed:
- added 'dataset' scores, mirroring 'model state' scores, which can be used as a separate or complimentary reward
- added 'sampling only' mode as option in al pipeline, to just run MCMC, random + annealing, or GFlowNet (partially tested - it does run but may be lingeringi issues around config + oracle) using a standard AL config. There are some unnecessary inits still but they don't hurt anything.

InfluenceFunctional committed:
- added extra seed to annealer and random sampler so that it doesn't pick the same RNG at every AL iter

InfluenceFunctional committed:
- added extra seed to annealer and random sampler so that it doesn't pick the same RNG at every AL iter

InfluenceFunctional committed:
- slight speedup in model evaluation
- added configs for pure sampling runs
- adde diversity measurement by agglomerative clustering to pure sampling

InfluenceFunctional committed:
- mcmc fix - missing variable name

InfluenceFunctional committed:
- mcmc fix - missing variable name

alexhernandezgarcia commented on gflownet.py:
```diff
@@ -41,7 +41,7 @@ def add_args(parser):
         "-y",
         "--yaml_config",
-        default=None,
+        default="config/mk_test_defaults_gfn_only.yml",
```
Do we prefer to have a specific yml as default here? (sorry if this was answered already and I missed it)

InfluenceFunctional replied:
no this was another lazy move by me - if anything it should be set to default.yml, which we have yet to create

alexhernandezgarcia replied:
Alright! I'll change back to None and open an issue to create the default.yml

alexhernandezgarcia committed:
change default yml config back to None

alexhernandezgarcia commented on gflownet.py:
```diff
@@ -403,8 +425,9 @@ def __init__(self, args, comet=None, proxy=None, al_iter=-1, data_path=None):
                     seed=args.gflownet.test.seed,
-                    output_csv=args.gflownet.test.output,
                 )
+                # output_csv=args.gflownet.test.output,
+                # )
```
I guess this is commented out because it was throwing an error?

InfluenceFunctional replied:
I think I was trying to get make_uniform_test_set to work - I actually never did so this is still 'open' for me

alexhernandezgarcia replied:
Will revert back to the original.

alexhernandezgarcia commented on gflownet.py:
```diff
@@ -1347,6 +1369,19 @@ def main(args):
     gflownet_agent = GFlowNetAgent(args)
     gflownet_agent.train()
 
+    # sample from the oracle, not from a proxy model
+    samples, times = gflownet_agent.sample(
+        args.gflownet.n_samples,
+        args.gflownet.max_seq_length,
+        args.gflownet.min_seq_length,
+        args.gflownet.nalphabet,
+        args.gflownet.min_word_len,
+        args.gflownet.max_word_len,
+        gflownet_agent.env.oracle,
+        mask_eos=True,
+        get_uncertainties=False,
+    )
+
```
What is this part for? What are `samples` and `times` used for?

InfluenceFunctional replied:
there was no sampling method when running gflownet on its own, so I copied this from elsewhere in the code. In principle one could/should save the samples for analysis

alexhernandezgarcia replied:
Yeah, those orphans look a bit orphan useless otherwise :P If you could write the saving part that would be great, but no high priority

alexhernandezgarcia committed:
revert commented lines

alexhernandezgarcia committed:
black main

alexhernandezgarcia committed:
add prefix gflownet_ to new gflownet args

alexhernandezgarcia commented on main.py:
```diff
@@ -549,6 +575,8 @@ def process_config(config):
         ]
         config.dataset.dict_size = 4
     # GFlowNet
+    config.gflownet.post_annealing_samples = config.al.annealing_samples
+    config.gflownet.post_annealing_time = config.al.annealing_time
```
I think the `config.gflownet.*` variables are never used and `config.al.*` are use instead (correctly). Shall we delete these lines?

InfluenceFunctional replied:
this was something I actually wanted to talk about - there are other duplicated configs like max and min length - seems like they could be consolidated. My bias is to put them in the overall AL config rather than be GFN-specific, but want to see what you think

alexhernandezgarcia replied:
I agree that anything that is common to more than one sampler should go to the overall config indeed. But `gflownet` still needs to get passed those variables that it depends on when run directly, which is the case of min and max length.

I will delete these lines for now and let's do the same whenever we find new such candidates.

alexhernandezgarcia replied:
In case I am not clear: we can delete `post_annealing_samples` and `post_annealing_time` because they are not used in `gflownet.py`, but `min_seq_length` is used there and it has to be `gflownet.min_seq_length` because when run directly there will not be `dataset.min_length`.

alexhernandezgarcia committed:
delete unnecessary lines

alexhernandezgarcia reviewed (approved):
Good stuff here!
